torchdata/datapipes/iter/util/plain_text_reader.py

```diff
@@ -0,0 +1,237 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+import csv
+from typing import Tuple, Union, Iterator, TypeVar
+
+from torch.utils.data import IterDataPipe, functional_datapipe
+
+D = TypeVar("D")
+
+
+class PlainTextReaderHelper:
+    def __init__(
+        self,
+        *,
+        skip_lines: int = 0,
+        strip_newline: bool = True,
+        decode: bool = True,
+        encoding="utf-8",
+        errors: str = "ignore",
+        return_path: bool = False,
+    ) -> None:
+        if skip_lines < 0:
+            raise ValueError("'skip_lines' is required to be a positive integer.")
+        self._skip_lines = skip_lines
+        self._strip_newline = strip_newline
+        self._decode = decode
+        self._encoding = encoding
+        self._errors = errors
+        self._return_path = return_path
+
+    def skip_lines(self, stream: Iterator[D]) -> Iterator[D]:
```

Comment:
The type hint seems not right for stream.
I have tried `issubclass(BinaryIO, Iterator)` and it returns False.

Reply:
Should we annotate the `source_datapipe` with `IterDataPipe[Tuple[str, io.BinaryIO]]`?

Reply:
Yes, please. I should also move the issue from torchdata that we need to add annotation to DataPipes.

Reply:
If I annotate with `BinaryIO`, `mypy` complains with

```
torchdata/datapipes/iter/util/plain_text_reader.py:31: error: The return type of a generator function should
be "Generator" or one of its supertypes  [misc]
        def skip_lines(self, stream: BinaryIO) -> BinaryIO:
```

Reply:
I mean the output can still be Iterator, right?

Reply:
```
torchdata/datapipes/iter/util/plain_text_reader.py:106:22: error: Incompatible types in assignment (expression
has type "Iterator[bytes]", variable has type "BinaryIO")  [assignment]
                stream = self._helper.skip_lines(stream)
```

Reply:
I don't think we can have both.

Reply:
Here is what I was thinking:
`def skip_lines(self, stream: IO) -> Iterator[D]:`
Using IO because it can be either `BinaryIO` or `TextIO`.

It requires developer must run `skip_lines` as the first operation to convert it from `IO` to `Iterator`. Then, the other methods can use `Iterator` as the input annotation.

Reply:
But what if we later add  a datapipe that doesn't use `skip_lines`. At this point we would be in trouble, no?

Reply:
Yeah, that's exact my concern.
We have two options:
- Add a method to convert `IO` to `Iterator` and always call it at the beginning
- Input stream annotation can be `Union[IO, Iterator]`

Reply:
I've gone with the initial suggestion to "always use `skip_lines` first". We can deal with the other cases as soon as we hit them, ok?